File: Solutions/261-270/264. 丑数 II[优先队列+哈希集合] 中等.cpp
/*
** Problem: https://leetcode-cn.com/problems/ugly-number-ii/
** Article: 
** Author: memcpy0
*/
class Solution {
public:
    int nthUglyNumber(int n) {
        using ll = long long;
        priority_queue<ll, vector<ll>, greater<ll>> pq;
        unordered_set<ll> st;
        pq.push(1);
        st.insert(1);
        int ps[3] = {2, 3, 5};
        while (n-- > 0) {
            ll u = pq.top(); pq.pop();
            if (n == 0) return (int)u;
            for (int i = 0; i < 3; ++i) {
                ll v = u * ps[i];
                if (st.find(v) == st.end()) { //there isn't this ugly number
                    pq.push(v);
                    st.insert(v);
                }
            }
        }
        return 0; //never
    }
};

File: Solutions/261-270/264. 丑数 II[优先队列+哈希集合] 中等.java
/*
** Problem: https://leetcode-cn.com/problems/ugly-number-ii/
** Article: 
** Author: memcpy0
*/
class Solution {
    public int nthUglyNumber(int n) {
        PriorityQueue<Long> pq = new PriorityQueue<>();
        Set<Long> st = new HashSet<>();
        pq.add(1L);
        st.add(1L);
        int[] ps = new int[]{2, 3, 5};
        for (int i = 1; i <= n; ++i) {
            long u = pq.poll();
            if (i == n) return (int)u;
            for (int p : ps) {
                long v = u * p;
                if (!st.contains(v)) {
                    pq.add(v);
                    st.add(v);
                }
            }
        }
        return 0;
    }
}